[DOC] Add transactions how-to and example (#4348)
pachyderm/pachyderm: doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md

Pachyderm logs to `stderr` to indicate that the command was placed
in a transaction rather than run directly.

## Multiple Opened Transactions

Some systems have a notion of *nested* transactions. That is when you
open transactions within an already opened transaction. In such systems, the
operations added to the subsequent transactions are not executed
until all the nested transactions and the main transaction are closed.

Pachyderm does not support such behavior. Instead, when you open a
transaction, the transaction ID is written to the Pachyderm configuration
file. If you begin another transaction while the first one is open, Pachyderm
suspends the first transaction and overwrites the transaction ID in the
configuration file. All operations that you add to the ensuing
transactions will be executed as soon as you close those
transactions. To resume the initial transaction, you need to run
`pachctl resume transaction`.

Because nested transactions are not supported, transactions cannot
conflict with each other. Every time you add a command to a transaction,
Pachyderm creates a blueprint of the commit and verifies that the

> command is valid.

Reviewer: I know this is merged already, but this isn't clear to me for a couple of reasons:

1) AFAICS, whether transactions can nest is inconsequential to whether transactions can conflict.
2) What happens if two different users are running operations in their respective transactions at the same time? How are conflicting changes resolved? (e.g. if both transactions are creating a repo with the same name?)

Author: @Tryneus can you comment on the above?

Reviewer 2: @ysimonson 

>  whether transactions can nest is inconsequential to whether transactions can conflict

Agreed - and while transactions can't necessarily conflict (for some definition of "conflict"), one transaction can invalidate another (e.g. by deleting a branch that the other transaction references).

> What happens if two different users are running operations in their respective transactions at the same time? How are conflicting changes resolved?

Transactions are at their core a batch of operations that are applied to the cluster state in one atomic operation.  Building a transaction over multiple commands has no effect on cluster state until the transaction is committed, although each command verifies that the entire batch is still valid by doing a dry run.

If two transactions create the same repo, they will both be valid until one transaction is committed, then the other transaction will get an error when attempting to commit it or add more commands into the transaction.

# Use Pachyderm Transactions

!!! note "Summary"
    Use transactions to run multiple Pachyderm commands
    simultaneously in one job run.

A transaction is a Pachyderm operation that enables you to create
a collection of Pachyderm commands and execute them concurrently in a
single job run. Regular Pachyderm operations are executed consequently,
one after another. However, when you need to run multiple commands
at the same time, you can use transactions.
The transaction framework provides a method for batching together
commit propagation such that changed branches are collected over
the course of the transaction and all propagated in one batch at
the end. This method allows Pachyderm to dedupe changed branches and
branches provenant on the changed branches so that the minimum

> number of new commits are issued.

Author: removed

```

When you finish the transaction, both repositories switch to
to the master at the same time which creates new commits in
the corresponding `master` branches and the pipeline triggers
one job to process the commits together.

## Start and Finish Transactions

To start a transaction, run the following command:

```bash
$ pachctl start transaction
Started new transaction: 7a81eab5-e6c6-430a-a5c0-1deb06852ca5
```

This command generates a transaction object in the cluster and saves
its ID in the local Pachyderm configuration file. By default, this file
is stored at `~/.pachyderm/config.json`.

!!! example
    ```json hl_lines="9"
    {
       "user_id": "b4fe4317-be21-4836-824f-6661c68b8fba",
       "v2": {
         "active_context": "local-2",
         "contexts": {
           "default": {},
           "local-2": {
             "source": 3,
             "active_transaction": "7a81eab5-e6c6-430a-a5c0-1deb06852ca5",
             "cluster_name": "minikube",
             "auth_info": "minikube",
             "namespace": "default"
           },
    ```

After you start a transaction, you can add supported commands, such

> as create a repo, create a branch, and so on, to th transaction. You

Reviewer 3: ```suggestion
as `create branch`, `start commit`, and so on, to the transaction. You
```

Reviewer 3: remove: 

"you can also start a commit within a transaction"

can also start a commit within a transaction. If you close a transaction
before you close a commit, the commands that you have specified inside
the transaction and inside the
commit will not be applied until you close all commits that you have

> started in that transaction.

Reviewer 3: > If you close a transaction
before you close a commit, the commands that you have specified inside
the transaction and inside the
commit will not be applied until you close all commits that you have
started in that transaction.

I dont really understand what this is saying and I dont think it's accurate. There's no special logic around waiting on commits or something with transactions. Commits have the property that they dont do anything until you close them. 

Perhaps a better way to describe this is:

"All commands that are performed in a transaction are queued up and not executed against the actual cluster. One you finish the transaction, all queued command are executed atomically."

| Command      | Description |
| ------------ | ----------- |
| `pachctl list transaction` | List all unfinished transactions available in the Pachyderm cluster. |
| `pachctl stop transaction` | Remove the currently active transaction from the local Pachyderm config file. The transaction remains in the Pachyderm cluster and can be resumed later. |
| `pachctl resume transaction` | Set an already-existing transaction as the active transaction in the local Pachyderm config file. |
| `pachctl delete transaction` | Deletes a transaction from the Pachyderm cluster. |
| `pachctl inspect transaction` | Provides detailed information about an existing transaction, including which operations it will perform. By default, displays information about the current transaction. If you specify a transaction ID, displays information
about the corresponding transaction. |

## Supported Operations

While there is a transaction object in the Pachyderm configuration
file, all supported API requests append the request to the
transaction instead of running directly. These supported commands include:

```bash
create repo
delete repo
start commit
finish commit
delete commit
create branch
delete branch
```

> Each time you ad a command to a transaction, Pachyderm validates the

Reviewer 3: ```suggestion
Each time you add a command to a transaction, Pachyderm validates the
```
